### PR TITLE
Tsmatrixlib

### PR DESCRIPTION
**Pull request by madguy02**

### CODE REVIEW

**Review by Sing-Li — CHANGES_REQUESTED** [review]
Have you tried running tests with this yet?  Can we add a couple of test to make sure it works.

**Comment by Sing-Li** [comment]
Please also move the readme over, as per our discussions.  And remove the old directory.

**Comment by madguy02** [comment]
@Sing-Li one set of test added to test_ts and transpilation automated with npm run start.

**Comment by Sing-Li** [comment]
Please add sections on README.  Let a new contributor know how to:

* check out your branch
* build the code
* run the tests

Please make sure your instructions actually work - by checking out your code on a *clean* machine, VM, or VPS.

File: tsmatrixserverlib/Readme.md [hunk 1]
```diff
@@ -17,5 +17,22 @@ We selected TypeScript over JavaScript because:
 
 The is work in progress.  We will update this page from time to time.
 
+#### Set up tsmatrixserverlib
+
+git clone<repository>.
```
Comment (Sing-Li) [inline tsmatrixserverlib/Readme.md]: `git clone` what?  Please be precise, for someone starting from scratch on a new system.

File: tsmatrixserverlib/Readme.md [hunk 2]
```diff
@@ -17,5 +17,22 @@ We selected TypeScript over JavaScript because:
...
+
+git checkout branchname(tsmatrixlib).
```
Comment (Sing-Li) [inline tsmatrixserverlib/Readme.md]: if it is going to be `tsmatrixlib`, then use `tsmatrixlib`.  This should be explicit instruction that works.

File: tsmatrixserverlib/Readme.md [hunk 3]
```diff
@@ -17,5 +17,22 @@ We selected TypeScript over JavaScript because:
...
+
+cd tsmatrixserverlib
+
+npm install
+
+npm run start
```
Comment (Sing-Li) [inline tsmatrixserverlib/Readme.md]: There is no server to "run".   You are creating a lib that is an npm module.  Ideally, it should build the module off an npm command.

**Review by Sing-Li — CHANGES_REQUESTED** [review]

File: tsmatrixserverlib/test_ts/hello_test.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import hello from '../index';
+import { expect } from 'chai';
+// if you used the '@types/mocha' method to install mocha type definitions, uncomment the following line
+ import 'mocha';
+
+describe('Hello function', () => {
```
Comment (Sing-Li) [inline tsmatrixserverlib/test_ts/hello_test.ts]: Let's not bring in "hello" test, and use some of the tests that you've written already.

The JS tests you've already written should pass, as well as new "type enforced" TS tests that you'll add.

**Review by Sing-Li — CHANGES_REQUESTED** [review]

**Review by Sing-Li — APPROVED** [review]

**Comment by Sing-Li** [comment]
`npm test` currently only run the TS tests.    

Can you please add instructions (workflow) on running the JS tests as well.

**Comment by Sing-Li** [comment]
`setup_XX.sh` actually **run** the tests , really should be  named `run_XX.sh` 

But right now, the JS one does not work `./setup_JS.sh: line 3: mocha: command not found`

We really should have `npm test` run both TS and JS tests, and not separate shell scripts that the users have to run.

**Comment by madguy02** [comment]
Okay I will correct it ASAP

**Comment by madguy02** [comment]
@Sing-Li  changes pushed

**Comment by madguy02** [comment]
@Sing-Li  please check whether this test directory stucture is proper

**Comment by Sing-Li** [comment]
@madguy02    any reason why you're preceeding with`test_` again?  these subdir are already under the `test` subdirectory.   Secondly, `npm test` currently does not work as of this commit.

**Comment by Sing-Li** [comment]
Excellent!   Let me know if you want a merge down for the interim.

**Comment by madguy02** [comment]
Yeah @Sing-Li .. a merge down will help,as I was planning to start another PR on the rest of the module development.

**Pull request merged**